Userland/Libraries/LibCore/Environment.cpp

```diff
@@ -0,0 +1,149 @@
+ */
+
+#include "Environment.h"
+#include <AK/ByteString.h>
+
+#if defined(AK_OS_MACOS)
+#    include <crt_externs.h>
+#else
+extern char** environ;
+#endif
+
+namespace Core::Environment {
+
+char** raw_environ()
+{
+#if defined(AK_OS_MACOS)
+    return *_NSGetEnviron();
+#else
+    return environ;
+#endif
+}
+
+static Entry entry_from_string(char const* raw_input)
+{
+    StringView input { raw_input, strlen(raw_input) };
+    auto split_index = input.find('=');
+    if (!split_index.has_value()) {
+        return Entry {
+            .full_entry = input,
+            .name = input,
+            .value = ""sv,
+        };
+    }
+
+    return Entry {
+        .full_entry = input,
+        .name = input.substring_view(0, *split_index),
+        .value = input.substring_view(*split_index + 1),
+    };
+}
+
+void for_each_entry(Function<IterationDecision(Entry&)> callback)
+{
+    auto environment = raw_environ();
+    for (size_t i = 0; environment[i]; ++i) {
+        auto entry = entry_from_string(environment[i]);
+        if (callback(entry) == IterationDecision::Break)
+            break;
+    }
+}
+
+size_t size()
+{
+    auto environment = raw_environ();
+
+    size_t environ_size = 0;
+    while (environment[environ_size])
+        ++environ_size;
+
+    return environ_size;
+}
+
+Vector<Entry> entries()
+{
+    Vector<Entry> entries;
+
+    for_each_entry([&entries](Entry& entry) {
+        entries.append(entry);
+        return IterationDecision::Continue;
+    });
+
+    return entries;
+}
+
+Optional<StringView> get(StringView name)
+{
+    auto* result = ::getenv(name.to_byte_string().characters());
+    if (result)
+        return StringView { result, strlen(result) };
+    return {};
+}
+
+Optional<StringView> get_secure(StringView name)
+{
+    auto* result = ::secure_getenv(name.to_byte_string().characters());
+    if (result)
+        return StringView { result, strlen(result) };
+    return {};
+}
+
+ErrorOr<void> set(StringView name, StringView value, Overwrite overwrite)
+{
+    auto builder = TRY(StringBuilder::create());
+    TRY(builder.try_append(name));
+    TRY(builder.try_append('\0'));
+    TRY(builder.try_append(value));
+    TRY(builder.try_append('\0'));
+    // Note the explicit null terminators above.
+    auto c_name = builder.string_view().characters_without_null_termination();
+    auto c_value = c_name + name.length() + 1;
+    auto rc = ::setenv(c_name, c_value, overwrite == Overwrite::Yes ? 1 : 0);
+    if (rc < 0)
+        return Error::from_errno(errno);
+    return {};
+}
+
+ErrorOr<void> unset(StringView name)
+{
+    auto builder = TRY(StringBuilder::create());
+    TRY(builder.try_append(name));
+    TRY(builder.try_append('\0'));
+
+    // Note the explicit null terminator above.
+    auto rc = ::unsetenv(builder.string_view().characters_without_null_termination());
+    if (rc < 0)
+        return Error::from_errno(errno);
+    return {};
+}
+
+ErrorOr<void> put(StringView env)
```

Comment:
Is there any reason to prefer this flawed `putenv` over `setenv`?

Reply:
:man_shrugging: I'm just matching the LibC API.

Reply:
You definitely don't have to, even POSIX says, "The setenv() function is preferred over this function."

The single use in SS/Service.cpp can be safely changed to setenv.